### PR TITLE
Add timeout queue

### LINKED ISSUE #1115: DOM memory leak?
**Current Behavior**  
A clear and concise description of what is happening / what the bug is.  
DOM memory climbs when you open a different dataset. 

**Expected behavior**  
A clear and concise description of what you expected to happen instead.  
DOM memory should be reclaimed when you are finished with a dataset.

**How to reproduce**  
Steps to reproduce the current behavior:  
1. Open the chrome task manager with Shift-esc 
2. make sure you have a column for memory footprint (this shows DOM memory)
3. http://localhost:4000/                                                  55k
4. http://localhost:4000/flu/seasonal/h1n1pdm/ha/12y 177k
5. http://localhost:4000/ 
6. http://localhost:4000/flu/seasonal/yam/ha/12y          232k dropping back to 166k
7. http://localhost:4000/                                                 170k
8. http://localhost:4000/flu/seasonal/h1n1pdm/na/6y    255k dropping back to 212k

Hi all,
Not bad, it recovers memory. Now rapidly choose a data set then use the back button
4 or 5 times. I am looking at 425k now, and it is not dropping back. Admittedly this is gorilla testing, it would not affect normal users. So let's call this a bug of minor importance. Moving on ..
thanks -- Rick

**Possible solution**  
(optional)  

**Your environment: if browsing Nextstrain online**  
 - Operating system:  
 - Browser:  

**Your environment: if running Nextstrain locally**  
 - Operating system:   Ubuntu 91.10
 - Browser:    Chrome
 - Version (e.g. `auspice 2.7.0`):   master  (2.14.0)

**Additional context**

### PR DESCRIPTION
**Pull request by dos077**
### Description of proposed changes    
Replace anonymous setTimeout in components with mapped timeout arrays that's traceable and clearable for house keeping at unmount.

### Related issue(s)  

Fixes #1115 

### Testing
Tested by navigating around datasets and new unit test for the util/timeoutqueue added.

### CODE REVIEW

**Comment by jameshadfield** [comment]
Exciting! Looking forward to testing this. I believe most of the timeouts only exist to make libraries play nicely together, indicating that they could be removed in an ideal world. I know I've used them in `<Tree>` because keeping track of thousands of individual d3 transitions proved to be a big performance hit.

File: src/components/map/map.js [hunk 1]
```diff
@@ -144,10 +145,12 @@ class Map extends React.Component {
     if (this.state.map && (this.props.width !== nextProps.width || this.props.height !== nextProps.height)) {
       // first, clear any existing timeout
       if (this.map_timeout) {
-        window.clearTimeout(this.map_timeout);
+        removeTimeout(this.map_timeout);
```
Comment (jameshadfield) [inline src/components/map/map.js:148]: needs to be `removeTimeout('map', this.map_timeout)`

File: src/components/map/map.js [hunk 2]
```diff
@@ -640,10 +643,12 @@ class Map extends React.Component {
     const maxZoom = this.getMaxZoomForFittingMapToData();
     // first, clear any existing timeout
     if (this.bounds_timeout) {
-      window.clearTimeout(this.bounds_timeout);
+      removeTimeout(this.bounds_timeout);
```
Comment (jameshadfield) [inline src/components/map/map.js:646]: needs to be `removeTimeout('map', this.map_timeout)`

File: src/components/tree/tangle/index.js [hunk 3]
```diff
@@ -72,11 +73,11 @@ class Tangle extends React.Component {
   componentDidUpdate(prevProps) {
     if (this.props.width !== prevProps.width || this.props.height !== prevProps.height) {
       if (this.timeout) {
-        window.clearTimeout(this.timeout);
+        removeTimeout(this.timeout);
```
Comment (jameshadfield) [inline src/components/tree/tangle/index.js:76]: needs to be removeTimeout('tree', this.timeout)

File: src/util/timeoutQueue.js [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+  const newTimeout = setTimeout(() => {
+    fn(...args);
+    timeouts[component] = timeouts[component].filter((to) => to !== newTimeout);
+  }, delay);
+  timeouts[component].push(newTimeout);
+  return newTimeout;
+};
+
+export const removeTimeout = (component, timeout) => {
+  if (timeouts[component].some((to) => to === timeout)) {
```
Comment (jameshadfield) [inline src/util/timeoutQueue.js:14]: Consider adding a safeguard here such as 
```js
  if (!timeouts[component]) {
    console.error("Cannot remove timeout for component", component, "ref", timeout);
    return;
  }
```
or wrapping with try/catch

File: src/components/map/map.js (hunk 2 above)
Comment (dos077) [reply src/components/map/map.js:646]: oops, sorry

**Comment by dos077** [comment]
Something gone terribly wrong when I merged upstream changes to my fork and rebase the feature branch... Gonna redo the pull request again...

**Pull request closed without merging**